lib/standard/stream.nit

```diff
@@ -405,66 +413,79 @@ abstract class BufferedReader
 		return c
 	end
 
-	# Peeks up to `n` bytes in the buffer, returns an empty string on EOF
+	# Peeks up to `n` bytes in the buffer
 	#
 	# The operation does not consume the buffer
-	#
-	# ~~~nitish
-	# 	var x = new FileReader("File.txt")
-	#	assert x.peek(5) == x.read(5)
-	# ~~~
-	fun peek(i: Int): String do
-		if eof then return ""
-		var b = new FlatBuffer.with_capacity(i)
-		while i > 0 and not eof do
-			b.add _buffer[_buffer_pos]
-			_buffer_pos += 1
-			i -= 1
+	fun peek(i: Int): Bytes do
+		if eof then return new Bytes.empty
+		var remsp = _buffer_length - _buffer_pos
+		if i <= remsp then
+			var bf = new Bytes.with_capacity(i)
+			bf.append_ns_from(_buffer, i, _buffer_pos)
+			return bf
 		end
-		var nbuflen = b.length + (_buffer.length - _buffer_pos)
-		var nbuf = new FlatBuffer.with_capacity(nbuflen)
-		nbuf.append(b)
-		while _buffer_pos < _buffer.length do
-			nbuf.add(_buffer[_buffer_pos])
-			_buffer_pos += 1
+		var bf = new Bytes.with_capacity(i)
+		bf.append_ns_from(_buffer, remsp, _buffer_pos)
+		_buffer_pos = _buffer_length
+		read_intern(i - bf.length, bf)
+		remsp = _buffer_length - _buffer_pos
+		var full_len = bf.length + remsp
+		if full_len > _buffer_capacity then
+			var c = _buffer_capacity
+			while c < full_len do c = c * 2 + 2
+			_buffer_capacity = c
 		end
+		var nns = new NativeString(_buffer_capacity)
+		bf.items.copy_to(nns, bf.length, 0, 0)
+		_buffer.copy_to(nns, remsp, _buffer_pos, bf.length)
+		_buffer = nns
 		_buffer_pos = 0
-		_buffer = nbuf
-		return b.to_s
+		_buffer_length = full_len
+		return bf
 	end
 
 	redef fun read(i)
 	do
-		if last_error != null then return ""
-		if eof then return ""
+		if last_error != null then return new Bytes.empty
+		var buf = new Bytes.with_capacity(i)
+		read_intern(i, buf)
+		return buf
+	end
+
+	# Fills the `buf` with the values read from the source
```

Comment:
If you don't use the word `buffer` I suggest to remove the "the". Also, I don't understand the `source` concept here.

A suggestion:
~~~nit
# Fills `buf` with `i` bytes read from `self`.
~~~